bootstrap/internal/controllers/rke2config_controller.go

```diff
@@ -334,14 +334,28 @@ func (r *RKE2ConfigReconciler) handleClusterNotInitialized(ctx context.Context,
 
 	conditions.MarkTrue(scope.Config, bootstrapv1.CertificatesAvailableCondition)
 
-	token, err := r.generateAndStoreToken(ctx, scope)
-	if err != nil {
-		scope.Logger.Error(err, "unable to generate and store an RKE2 server token")
+	// RKE2 server token must only be generated once, so all nodes join the cluster with the same registration token.
+	var token string
 
-		return ctrl.Result{}, err
+	tokenSecret := &corev1.Secret{}
+	secretKey := types.NamespacedName{
+		Namespace: scope.Config.Namespace,
+		Name:      bsutil.TokenName(scope.Cluster.Name),
 	}
+	err := r.Client.Get(ctx, secretKey, tokenSecret)
```

Comment:
I'm under the impression that this code might be subject to a race condition ; for instance in the case where we would run multiple replicas of the controller (both controller could fail the get in the same window of time before any reaches the create, and then both would do the create).

I think that the right pattern here would be to:
- try to create
- if this fails because the object already exists, then read it's content

Maybe one possibility is to make this update step of createOrUpdateSecretFromObject optional:
https://github.com/rancher-sandbox/cluster-api-provider-rke2/blob/main/bootstrap/internal/controllers/rke2config_controller.go#L827-L829

and call createOrUpdateSecretFromObject from generateAndStoreToken with this update step disabled

Reply:
Thanks for the feedback @tmmorin. My initial idea was to create a separate `createSecretFromObject` and call this function to generate the token, so that the update functionality is not altered and we don't have to deal with the value being updated, but decided to opt for this simpler solution. Didn't think about the corner case you described. I'll review this and take the other approach.

Reply:
Thanks for solving this problem.